runtime/runtime/src/lib.rs

```diff
@@ -319,12 +403,13 @@ impl Runtime {
                     block_index,
                     nonce.as_ref().to_vec(),
                     false,
+                    public_key,
                 ),
             )
             .map_err(|e| format!("wasm async call preparation failed with error: {:?}", e))?;
             transaction_result.logs.append(&mut wasm_res.logs);
             let balance = wasm_res.frozen_balance;
-            *leftover_balance += wasm_res.liquid_balance;
+            *leftover_balance = wasm_res.liquid_balance;
```

Comment:
Before this change if wasm execution fails , e.g. due to running out of liquid balance, the leftover_balance equals 0, because we lose all liquid balance. With this change if execution fails the leftover balance is equal to the entire liquid balance that was before the execution started, which creates an abuse vector since users can grind validators by sending transactions that fail and get all liquid balance back through the refund.

Reply:
Contract can fail by calling `assert` internally. That's why this entire logic has to be changed to account for balances correctly, like it was before. Which means execution runtime should not be able to remove exceed the initial liquid balance, and frozen balance withdrawals should not modify the account balance in case of contract failure

Reply:
Is this the spec that you are currently suggesting?
* We use liquid balance for: wasm execution, fee for creating initial transaction;
* We use frozen balance for: promise creation and other transaction creation within contract, rent;

Failure modes:
* When contract fails due to running out of liquid balance the liquid balance is lost and frozen balance is reverted to value before the execution;
* When contract fails due to panic (assert or out of bounds) the same rules applies.

Correct?

Reply:
Correct. But we also need to account for deposited liquid balance. In case of failure we need to return it back to the liquid balance (since we reverting frozen balance) and the rest of the liquid balance has to be refunded